### PR TITLE
CertMan Automated Reissuance Loop

### PR DESCRIPTION
**Pull request by DanielSollis**
### Scope of changes

This PR implements a new loop in the Certificate Manager which handles automated certificate reissuance.

### Type of change

- [ ] bug fix
- [x] new feature
- [ ] documentation
- [ ] other (describe)

### Acceptance criteria

Unit tests should be passing.

### Author checklist

- [x] I have manually tested the change and/or added automation in the form of unit tests or integration tests
- [ ]  I have updated the dependencies list
- [ ]  I have recompiled and included new protocol buffers to reflect changes I made
- [ ]  I have added new test fixtures as needed to support added tests
- [x]   Check this box if a reviewer can merge this pull request after approval (leave it unchecked if you want to do it yourself)
- [x]  I have moved the associated Shortcut story to "Ready for Review"

### Reviewer(s) checklist

- [ ] Any new user-facing content that has been added for this PR has been QA'ed to ensure correct grammar, spelling, and understandability.

### CODE REVIEW

File: pkg/gds/config/config.go [hunk 1]
```diff
@@ -88,9 +88,10 @@ type EmailConfig struct {
 }
 
 type CertManConfig struct {
-	Interval time.Duration `split_words:"true" default:"10m"`
-	Storage  string        `split_words:"true" required:"false"`
-	Sectigo  sectigo.Config
+	RequestInterval    time.Duration `split_words:"true" default:"10m"`
+	ReissuenceInterval time.Duration `split_words:"true" default:"24h"`
```
Comment (pdeziel) [inline pkg/gds/config/config.go:92]:
```suggestion
	ReissuanceInterval time.Duration `split_words:"true" default:"24h"`
```

Note that this will have to be updated in the other files as well.

File: pkg/gds/config/config_test.go [hunk 2]
```diff
@@ -124,7 +125,7 @@ func TestConfig(t *testing.T) {
 	require.Equal(t, testEnv["GDS_EMAIL_STORAGE"], conf.Email.Storage)
 	require.True(t, conf.Email.Testing)
 	require.Equal(t, testEnv["GDS_DIRECTORY_ID"], conf.Email.DirectoryID)
-	require.Equal(t, 1*time.Minute, conf.CertMan.Interval)
+	require.Equal(t, 1*time.Minute, conf.CertMan.RequestInterval)
```
Comment (pdeziel) [inline pkg/gds/config/config_test.go:128]: And also check that the reissuance interval config value is set?

File: pkg/gds/certman/certs.go [hunk 3]
```diff
@@ -675,3 +680,147 @@ func (c *CertificateManager) getCertStorage() (path string, err error) {
...
+
+		switch daysBeforeReissuance := timeBeforeExpiration.Hours() / 24; {
+		// Seven days before expiration, send a cert reissuance reminder to VASP.
+		case daysBeforeReissuance <= 7:
+			c.email.SendContactReissuanceReminder(vasp, 7, reissuanceDate)
+
+		// Ten days before expiration, reissue the VASP's identity certificate, send the email with the created pkcs12
+		// password and send the whisper link, as well as notifying the TRISA admin that reissuance has started.
+		case daysBeforeReissuance <= 10:
+			c.reissueIdentityCertificates(vasp)
```
Comment (pdeziel) [inline pkg/gds/certman/certs.go:719]: As the PR currently stands, I don't think anything is preventing the reissuance process from triggering multiple times before the other certman loop handles it. I think we need to check if there's an existing certificate request in `READY_TO_SUBMIT` on the VASP to avoid the double reissue. If we need to tackle this in a separate story that's fine, since it probably requires writing a helper method to do the check.

File: pkg/gds/emails/client.go [hunk 4]
```diff
@@ -62,6 +63,10 @@ type EmailManager struct {
 	adminsEmail  *mail.Address
 }
 
+func (m *EmailManager) ServiceEmail() *mail.Address {
+	return m.serviceEmail
+}
```
Comment (pdeziel) [inline pkg/gds/emails/client.go:68]: Can you check if this accessor is still needed?

File: pkg/gds/emails/client.go [hunk 5]
```diff
@@ -366,6 +371,91 @@ func (m *EmailManager) SendExpiresAdminNotification(vasp *pb.VASP, reissueDate t
...
+		VID:                 vasp.Id,
+		CommonName:          vasp.CommonName,
+		Endpoint:            vasp.TrisaEndpoint,
+		RegisteredDirectory: m.conf.ServiceEmail,
+		Reissuance:          reissuanceDate,
+	}
+
+	if vasp.IdentityCertificate != nil {
+		ReissuanceData.SerialNumber = strings.ToUpper(hex.EncodeToString(vasp.IdentityCertificate.SerialNumber))
+		ReissuanceData.Expiration, _ = time.Parse(time.RFC3339, vasp.IdentityCertificate.NotAfter)
```
Comment (pdeziel) [inline pkg/gds/emails/client.go:395]: Let's handle the parsing error here.

File: pkg/gds/emails/mock.go [hunk 6]
```diff
@@ -139,7 +139,7 @@ func (c *mockSendGridClient) Send(msg *sgmail.SGMailV3) (rep *rest.Response, err
 		}, err
 	}
 
-	// Email recepients are stored in Personalizations
+	// Email recipients are stored in Personalizations
```
Comment (pdeziel) [inline pkg/gds/emails/mock.go:142]: Thanks for correcting my spelling mistakes!

File: pkg/gds/models/v1/models.go [hunk 7]
```diff
@@ -172,6 +172,28 @@ func AppendAdminEmailLog(vasp *pb.VASP, reason string, subject string) (err erro
...
+
+		matchedReason := reason == value.Reason
+		withinTimeWindow := timestamp.After(time.Now().AddDate(0, 0, -timeWindowDays))
+
+		if matchedReason && withinTimeWindow {
+			sent++
+		}
+	}
+	return sent, nil
+}
```
Comment (pdeziel) [inline pkg/gds/models/v1/models.go:195]: There's some duplicate code between this method and the one in `contacts.go`, however it's fine if we have to open a separate story to do the cleanup.

File: pkg/gds/emails/client.go [hunk 8]
```diff
@@ -366,6 +371,91 @@ func (m *EmailManager) SendExpiresAdminNotification(vasp *pb.VASP, reissueDate t
...
+	}
+
+	if serviceEmail, err = mail.ParseAddress(m.serviceEmail.Address); err != nil {
+		log.Error().Err(err).Msg("could not parse the configured service email address")
+		return
+	}
```
Comment (pdeziel) [inline pkg/gds/emails/client.go:401]: Maybe I'm missing something but don't we already have the `*mail.Address` on the certman struct?

File: pkg/gds/emails/client.go [hunk 9]
```diff
@@ -366,6 +371,91 @@ func (m *EmailManager) SendExpiresAdminNotification(vasp *pb.VASP, reissueDate t
...
+			if err = models.AppendEmailLog(contact, reissuanceReminder, msg.Subject); err != nil {
+				log.Error().Err(err).Msg(fmt.Sprintf("error appending to %s's email log", contact))
+			}
+			verifiedContacts++
+		case "":
+			if verifiedContacts == 0 {
+				log.WithLevel(zerolog.FatalLevel).Msg(fmt.Sprintf("cert-manager could not find a verified contact for %s", vasp.Id))
+			}
+		}
+	}
```
Comment (pdeziel) [inline pkg/gds/emails/client.go:456]: Can we check the `iter.Error()` here and handle appropriately?

File: pkg/gds/emails/client.go [hunk 10]
```diff
@@ -366,6 +371,91 @@ func (m *EmailManager) SendExpiresAdminNotification(vasp *pb.VASP, reissueDate t
...
+			}
+		case "legal", "billing":
+			if err = m.Send(msg); err != nil {
+				log.Error().Err(err).Msg(fmt.Sprintf("error sending %s's %s reissuance reminder email", vasp.Id, contact))
+			}
+			if err = models.AppendEmailLog(contact, reissuanceReminder, msg.Subject); err != nil {
+				log.Error().Err(err).Msg(fmt.Sprintf("error appending to %s's email log", contact))
+			}
+			verifiedContacts++
+		case "":
```
Comment (pdeziel) [inline pkg/gds/emails/client.go:451]: Can we change this to the default case to capture any unexpected strings? Also, we do have constants defined in the `models` package for the technical, administrative, legal, billing strings.

File: pkg/gds/emails/client.go [hunk 11]
```diff
@@ -366,6 +371,91 @@ func (m *EmailManager) SendExpiresAdminNotification(vasp *pb.VASP, reissueDate t
...
+			ReissuanceData,
+		)
+		if err != nil {
+			log.Error().Err(err).Msg(fmt.Sprintf("could not create %s's %s reissuance reminder email", vasp.Id, contact))
+		}
+
+		// Send the reissuance reminder email to the VASP's verified contacts using the following logic:
+		// 		1. Send to the Technical contact if verified, else
+		// 		2. Send to the Administrative contact if verified, else
+		// 		3. Send to all other verified contacts
```
Comment (pdeziel) [inline pkg/gds/emails/client.go:431]: This logic is a bit hard to follow for me with the switch cases and the loop, and I'm not sure if it's matching what's described in the comment. Do you think it would be easier to break this into two loops? For example, the first loop could just build the list of verified contacts to send the emails to, and the second loop would simply send the emails to all contacts in the built list.

**Review by pdeziel — COMMENTED** [review]
@DanielSollis This is a lot to go over but I like the overall structure a lot and I think we just need to add some error handling and some slight refactoring to improve the readability.

File: pkg/gds/certman/certs.go [hunk 12]
```diff
@@ -111,20 +113,23 @@ func (c *CertificateManager) CertManager(stop <-chan struct{}) {
 
 	// Ticker is created in the go routine to prevent backpressure if the cert manager
 	// process takes longer than the specified ticker interval.
-	ticker := time.NewTicker(c.conf.Interval)
-	log.Info().Dur("interval", c.conf.Interval).Str("store", certDir).Msg("cert-manager process started")
+	requestsTicker := time.NewTicker(c.conf.RequestInterval)
+	reissuanceTicker := time.NewTicker(c.conf.ReissuenceInterval)
+	log.Info().Dur("interval", c.conf.RequestInterval).Str("store", certDir).Msg("cert-manager process started")
```
[the remark this answers is not in the document]
Comment (DanielSollis) [reply pkg/gds/certman/certs.go]: good idea, Fixed!

File: pkg/gds/certman/certs.go [hunk 13]
```diff
@@ -675,3 +680,147 @@ func (c *CertificateManager) getCertStorage() (path string, err error) {
 	log.Warn().Str("certs", path).Msg("using a temporary directory for cert downloads")
 	return path, err
 }
+
+// HandleCertificateReissuance iterates through each VASP in the database and checks
+// if their identity certificate will be expiring soon, sending reminder email within
```
[the remark this answers is not in the document]
Comment (DanielSollis) [reply pkg/gds/certman/certs.go]: nice catch, I updated the comment.

File: pkg/gds/certman/certs.go [hunk 14]
```diff
@@ -675,3 +680,147 @@ func (c *CertificateManager) getCertStorage() (path string, err error) {
...
+		err            error
+		expirationDate time.Time
+	)
+
+	// Iterate through the VASPs in the database.
+	vasps := c.db.ListVASPs()
+	for vasps.Next() {
+		var vasp *pb.VASP
+		if vasp, err = vasps.VASP(); err != nil {
+			log.Error().Err(err).Msg("could not parse vasp request from database")
```
[the remark this answers is not in the document]
Comment (DanielSollis) [reply pkg/gds/certman/certs.go]: comment fixed.

File: pkg/gds/emails/client.go (hunk 4 above)
Comment (DanielSollis) [reply pkg/gds/emails/client.go]: you're right, it's no longer needed, I deleted it.

File: pkg/gds/emails/mock.go (hunk 6 above)
Comment (DanielSollis) [reply pkg/gds/emails/mock.go:142]: Code spell checker is super helpful! https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker

File: pkg/gds/emails/client.go (hunk 9 above)
Comment (DanielSollis) [reply pkg/gds/emails/client.go:453]: checked and handled!

File: pkg/gds/emails/client.go (hunk 8 above)
Comment (DanielSollis) [reply pkg/gds/emails/client.go:393]: you're right, this is an artifact from when this was in a different package. Fixed!

File: pkg/gds/emails/client.go (hunk 5 above)
Comment (DanielSollis) [reply pkg/gds/emails/client.go]: parsing handled!

File: pkg/gds/certman/certs.go [hunk 15]
```diff
@@ -675,3 +680,147 @@ func (c *CertificateManager) getCertStorage() (path string, err error) {
...
+		log.Error().Err(err).Msg(fmt.Sprintf("error creating certificate request for vasp %s", vasp.Id))
+		return
+	}
+
+	// Update the cert req to be ready for submission.
+	if err = models.UpdateCertificateRequestStatus(
+		certreq,
+		models.CertificateRequestState_READY_TO_SUBMIT,
+		"automated certificate reissuance",
+		"support@rotational.io",
```
[the remark this answers is not in the document]
Comment (DanielSollis) [reply pkg/gds/certman/certs.go]: changed

File: pkg/gds/certman/certs.go [hunk 16]
```diff
@@ -675,3 +680,147 @@ func (c *CertificateManager) getCertStorage() (path string, err error) {
...
+		case daysBeforeReissuance <= 30:
+			c.email.SendContactReissuanceReminder(vasp, 30, reissuanceDate)
+			if NoAdminEmailSent(vasp, 30, reissuanceDate) {
+				if _, err = c.email.SendExpiresAdminNotification(vasp, reissuanceDate); err != nil {
+					log.Error().Err(err).Msg(fmt.Sprintf("error sending admin reissuance reminder for %s", vasp.Id))
+				}
+			}
+		}
+
+		c.db.UpdateVASP(vasp)
```
[the remark this answers is not in the document]
Comment (DanielSollis) [reply pkg/gds/certman/certs.go]: error handled!

File: pkg/gds/certman/certs.go [hunk 17]
```diff
@@ -675,3 +680,147 @@ func (c *CertificateManager) getCertStorage() (path string, err error) {
...
+		// Calculate the number of days before the VASP's certificate expires.
+		notAfter := vasp.IdentityCertificate.NotAfter
+		if expirationDate, err = time.Parse(time.RFC3339, notAfter); err != nil {
+			log.Error().Err(err).Msg(fmt.Sprintf("could not parse %s's cert reissuance date", vasp.Id))
+		}
+		reissuanceDate := expirationDate.Add(-time.Hour * 240) // Calculate the reissuance date, 10 days before expiration
+		timeBeforeExpiration := time.Until(expirationDate)
+
+		switch daysBeforeReissuance := timeBeforeExpiration.Hours() / 24; {
+		// Seven days before expiration, send a cert reissuance reminder to VASP.
```
[the remark this answers is not in the document]
Comment (DanielSollis) [reply pkg/gds/certman/certs.go:726]: comment augmented and variable name changed.

File: pkg/gds/certman/certs.go [hunk 18]
```diff
@@ -675,3 +680,147 @@ func (c *CertificateManager) getCertStorage() (path string, err error) {
...
+// the 30 and 7 day checkpoints if so, and reissuing the identity certificate 10 days before
+// expiration.
+func (c *CertificateManager) HandleCertificateReissuance() {
+	var (
+		err            error
+		expirationDate time.Time
+	)
+
+	// Iterate through the VASPs in the database.
+	vasps := c.db.ListVASPs()
```
[the remark this answers is not in the document]
Comment (DanielSollis) [reply pkg/gds/certman/certs.go:695]: Thanks, I didn't realize that, that's good to know, fixed!

File: pkg/gds/certman/certs.go [hunk 19]
```diff
@@ -675,3 +680,147 @@ func (c *CertificateManager) getCertStorage() (path string, err error) {
...
+		var vasp *pb.VASP
+		if vasp, err = vasps.VASP(); err != nil {
+			log.Error().Err(err).Msg("could not parse vasp request from database")
+			continue
+		}
+
+		// Calculate the number of days before the VASP's certificate expires.
+		notAfter := vasp.IdentityCertificate.NotAfter
+		if expirationDate, err = time.Parse(time.RFC3339, notAfter); err != nil {
+			log.Error().Err(err).Msg(fmt.Sprintf("could not parse %s's cert reissuance date", vasp.Id))
```
[the remark this answers is not in the document]
Comment (DanielSollis) [reply pkg/gds/certman/certs.go]: That's much better, I made the change to the log statement here and throughout the rest of the PR where applicable.

File: pkg/gds/certman/certs.go [hunk 20]
```diff
@@ -675,3 +680,147 @@ func (c *CertificateManager) getCertStorage() (path string, err error) {
...
+		}
+		reissuanceDate := expirationDate.Add(-time.Hour * 240) // Calculate the reissuance date, 10 days before expiration
+		timeBeforeExpiration := time.Until(expirationDate)
+
+		switch daysBeforeReissuance := timeBeforeExpiration.Hours() / 24; {
+		// Seven days before expiration, send a cert reissuance reminder to VASP.
+		case daysBeforeReissuance <= 7:
+			c.email.SendContactReissuanceReminder(vasp, 7, reissuanceDate)
```
[the remark this answers is not in the document]
Comment (DanielSollis) [reply pkg/gds/certman/certs.go]: changed!

File: pkg/gds/certman/certs.go [hunk 21]
```diff
@@ -675,3 +680,147 @@ func (c *CertificateManager) getCertStorage() (path string, err error) {
...
+			}
+
+		// Thirty days before expiration, send the reissuance reminder to the VASP and the TRISA admin.
+		case daysBeforeReissuance <= 30:
+			c.email.SendContactReissuanceReminder(vasp, 30, reissuanceDate)
+			if NoAdminEmailSent(vasp, 30, reissuanceDate) {
+				if _, err = c.email.SendExpiresAdminNotification(vasp, reissuanceDate); err != nil {
+					log.Error().Err(err).Msg(fmt.Sprintf("error sending admin reissuance reminder for %s", vasp.Id))
+				}
+			}
```
[the remark this answers is not in the document]
Comment (DanielSollis) [reply pkg/gds/certman/certs.go:754]: I created that story and added a TODO in the code: https://app.shortcut.com/rotational/story/8652

File: pkg/gds/certman/certs.go [hunk 22]
```diff
@@ -675,3 +680,147 @@ func (c *CertificateManager) getCertStorage() (path string, err error) {
...
+
+	// Iterate through the VASPs in the database.
+	vasps := c.db.ListVASPs()
+	for vasps.Next() {
+		var vasp *pb.VASP
+		if vasp, err = vasps.VASP(); err != nil {
+			log.Error().Err(err).Msg("could not parse vasp request from database")
+			continue
+		}
+
```
[the remark this answers is not in the document]
Comment (DanielSollis) [reply pkg/gds/certman/certs.go:715]: Check added

File: pkg/gds/config/config.go (hunk 1 above)
Comment (DanielSollis) [reply pkg/gds/config/config.go:92]: confirmed

File: pkg/gds/certman/certs.go (hunk 3 above)
Comment (DanielSollis) [reply pkg/gds/certman/certs.go:738]: I created a story for this in the automated reissuance epic.

File: pkg/gds/emails/client.go (hunk 11 above)
Comment (DanielSollis) [reply pkg/gds/emails/client.go]: I refactored this code and I think it is now significantly easier to follow

File: pkg/gds/certman/certs_test.go [hunk 23]
```diff
@@ -205,6 +205,166 @@ func (s *certTestSuite) TestCertManager() {
...
 }
 
+func (s *certTestSuite) TestCertManagerThirtyDayReissuanceReminder() {
+	_ = s.setupCertManager(sectigo.ProfileCipherTraceEE)
+	defer s.teardownCertManager()
+	defer s.fixtures.LoadReferenceFixtures()
+	require := s.Require()
+
+	// setup the datastore to contain the modified echoVASP
+	s.resetVASPDatabase()
```
Comment (pdeziel) [inline pkg/gds/certman/certs_test.go:215]: You wouldn't have known this because you haven't had to deal with the GDS fixtures, but our strategy throughout the GDS tests is that we rely on a set of "source of truth" VASP and certreq fixtures which are compressed into an archived file and included in the repo (at pkg/gds/testdata/fakes.tgz). Therefore, we generally prefer to use the existing fixtures rather than hacking our own inside of the tests (although sometimes we do have to modify them at test time). My strategy for writing tests that require fixtures is to check if there are any existing fixtures that are in the required state and if not, modify as necessary in the tests.

For this specific situation, I believe the requirements are that we need a verified VASP with verified administrative and technical contacts? In that case, it looks like `hotel` is a good fit, although it looks like you still need to set the administrative and technical contacts to something verified for your tests.

My suggestion is to use that VASP fixture and remove `resetVASPDatabase` because `teardownCertManager` will take care of that.

File: pkg/gds/certman/certs_test.go [hunk 24]
```diff
@@ -205,6 +205,166 @@ func (s *certTestSuite) TestCertManager() {
...
+		},
+		{
+			To:        s.conf.Email.AdminEmail,
+			From:      s.conf.Email.ServiceEmail,
+			Subject:   emails.ReissuanceAdminNotificationRE,
+			Reason:    "reissuance_admin_notification",
+			Timestamp: callTime,
+		},
+	}
+	emails.CheckEmails(s.T(), messages)
```
Comment (pdeziel) [inline pkg/gds/certman/certs_test.go:329]: Can we add a TODO here to test the email send logic? (e.g., if technical contact is not verified, email should be sent to the admin contact etc.)

File: pkg/gds/certman/certs_test.go [hunk 25]
```diff
@@ -205,6 +205,166 @@ func (s *certTestSuite) TestCertManager() {
...
+	// reissue the VASP's identity certificate, send the
+	// email with the created pkcs12 password and send
+	// the whisper link, as well as notifying the TRISA
+	// admin that reissuance has started.
+	s.updateVaspIdentityCert(echoVASP, 8)
+	callTime := time.Now()
+	s.certman.HandleCertificateReissuance()
+
+	v, err := s.db.RetrieveVASP(echoVASP.Id)
+	require.NoError(err)
```
Comment (pdeziel) [inline pkg/gds/certman/certs_test.go:308]: Since this starts the reissuance process, I'd also like to add a few more asserts here:

1. The certificate request is created and the state is in `READY_TO_SUBMIT`
2. A secret has been created in the secret manager
3. After two calls of `HandleCertificateRequests`, the VASP contains a new identity certificate, the cert request is completed, and there is a new certificate record in the database. (this might be too complicated for this PR so I'm fine with creating a follow up story for this)

Let me know if you need any help with the first two points.

**Review by pdeziel — COMMENTED** [review]
This is looking good, I just had a few comments regarding the fixture management in the testing and some small refactoring items.

File: pkg/gds/certman/certs_test.go (hunk 23 above)
Comment (DanielSollis) [reply pkg/gds/certman/certs_test.go:215]: Looking through the `fakes.tgz` file, I can't find any VASP fixtures that had both a verified administrative and technical contact. On `hotel` the technical contact is verified but the administrative contact is null.

Additionally, `resetVASPDatabase` is removing the superfluous VASPs from the database because `HandleCertificateReissuance` is using `db.ListVASPs` which means we would have to track the sent emails of the 10+ VASPs already in the database and all of their contacts.

File: pkg/gds/certman/certs_test.go [hunk 26]
```diff
@@ -682,7 +842,7 @@ func (s *certTestSuite) TestProcessBatchNoSuccess() {
 func (s *certTestSuite) TestCertManagerLoop() {
 	s.setupCertManager(sectigo.ProfileCipherTraceEE)
 	defer s.teardownCertManager()
-	s.runCertManager(s.conf.CertMan.Interval)
+	s.runCertManager(s.conf.CertMan.RequestInterval)
```
[the remark this answers is not in the document]
Comment (DanielSollis) [reply pkg/gds/certman/certs_test.go:904]: From what I can tell this test already covers that, but just to make sure I added a mirrored test and set s.conf.CertMan.ReissuanceInterval to 2ms.

File: pkg/gds/emails/mock.go [hunk 27]
```diff
@@ -241,12 +242,12 @@ func CheckEmails(t *testing.T, messages []*EmailMeta) {
 		sentEmails = append(sentEmails, msg)
 	}
 
-	for _, msg := range messages {
```
Comment (DanielSollis) [inline pkg/gds/emails/mock.go:244]: @pdeziel These changes were necessary because the current code assumed there would only be one email sent to each contact per test, which isn't the case with the `TestCertManagerReissuance` test. I added a TODO to refactor this code in the future.

File: pkg/gds/certman/certs_test.go [hunk 28]
```diff
@@ -205,9 +205,227 @@ func (s *certTestSuite) TestCertManager() {
...
+func (s *certTestSuite) TestCertManagerThirtyDayReissuanceReminder() {
+	_ = s.setupCertManager(sectigo.ProfileCipherTraceEE, fixtures.Small)
+	defer s.teardownCertManager()
+	defer s.fixtures.LoadReferenceFixtures()
+	require := s.Require()
+
+	// setup the datastore to contain the modified charlieVASP
+	charlieVASP, err := s.fixtures.GetVASP("charliebank")
+	require.NoError(err, "could not get charlie VASP")
+	//s.resetVASPDatabase(echoVASP)
```
Comment (pdeziel) [inline pkg/gds/certman/certs_test.go:217]:
```suggestion
```

File: pkg/gds/certman/certs_test.go [hunk 29]
```diff
@@ -205,9 +205,227 @@ func (s *certTestSuite) TestCertManager() {
...
+	charlieVASP = s.setupVASP(charlieVASP)
+
+	// Set the second VASP in the fixtures.Small set's verification status
+	// to REJECTED so that it does not get triggered for reissuance.
+	deltaVASP, err := s.fixtures.GetVASP("delta")
+	require.NoError(err)
+	deltaVASP.VerificationStatus = pb.VerificationState_REJECTED
+	require.NoError(s.db.UpdateVASP(deltaVASP))
+
+	// Capture the number of certificate requests on the delta VASP
```
Comment (pdeziel) [inline pkg/gds/certman/certs_test.go:303]:
```suggestion
	// Capture the number of certificate requests on the charlie VASP
```

File: pkg/gds/certman/certs_test.go [hunk 30]
```diff
@@ -205,9 +205,227 @@ func (s *certTestSuite) TestCertManager() {
...
+	require.NotEmpty(idCert.PublicKeyAlgorithm)
+	require.NotNil(idCert.Subject)
+	require.NotNil(idCert.Issuer)
+	_, err = time.Parse(time.RFC3339, idCert.NotBefore)
+	require.NoError(err)
+	_, err = time.Parse(time.RFC3339, idCert.NotAfter)
+	require.NoError(err)
+	require.False(idCert.Revoked)
+	require.NotEmpty(idCert.Data)
+	require.NotEmpty(idCert.Chain)
```
Comment (pdeziel) [inline pkg/gds/certman/certs_test.go:369]: Some additional test cases here would be to check if the VASP ended up in the right state `VERIFIED` and the new certificate request ended up in the right state `COMPLETED`. Those checks should be pretty easy to add, but I won't make it contingent for merging this PR since you added the deliver certs email check already.

File: pkg/gds/emails/mock.go (hunk 27 above)
Comment (pdeziel) [reply pkg/gds/emails/mock.go:244]: I see, should we change that to a greater or equal check instead?

**Review by pdeziel — COMMENTED** [review]
Thanks so much for bearing with my comments but I'm a lot more confident in the functionality now that we've improved the coverage and squashed the testing bugs.

There is just one more critical change I'd like to make; looks like it didn't get captured in this PR but it's really simple. On Line 92 of certs.go we need to remove the `go` from the call to `CertManager`. This was my mistake when I originally ported this to the certman package; the code as written starts a go routine within a go routine which means that the `wg.Done` gets called immediately instead of blocking on the `CertManager` call.

Otherwise very good job on this one and I mainly had a few cosmetic suggestions below.

File: pkg/gds/emails/mock.go (hunk 27 above)
Comment (DanielSollis) [reply pkg/gds/emails/mock.go:244]: I changed it to a GTE check but I let the TODO there, I think we could find a more elegant to handle this in the future.

**Review by pdeziel — APPROVED** [review]
Thanks for making the changes!
